examples/autoprog.py

```diff
@@ -0,0 +1,196 @@
+    target_message = "Target device ip address and port " + str(g_target_ip_addr) + " " + str(g_autoprogram_port)
+    mylogger.log_info(target_message)
+
+    spaceTargetMonitor(host=g_target_ip_addr, port=g_autoprogram_port)
+
+    mylogger.log_info("Initializing session to target ")
+    session = Session(
+        target=Target(
+            connection=TCPSocketConnection(host=g_target_ip_addr, port=g_autoprogram_port),
+            monitors=[spaceTargetMonitor],
+        ),
+        sleep_time=12,
+        crash_threshold_request=2,
+        crash_threshold_element=2,
+        index_start=int(start_index),
+        index_end=int(end_index),
+    )
+    define_autoprog_static(session=session)
+    mylogger.log_info("start fuzzing")
+
+    session.fuzz()
+
+
+#################################################################
+# Single entry configuration within a Proposal list:
+#################################################################
+# Example protocol which will send the following data in sequence:
+#    1. A URL inside of a formatted line
+#    2. The checksum of line 1 above.
+#    3. A formatted line to indicate the completion of the checksum.
+#    4. A formatted line indicating the start of the outer checksum.
+#    5. An outer checksum managing data consisting of lines 1,2,3, and 4.
+#    6. A formatted line to indicate the completion of the outer checksum.
+#
+# <ProposalList xmlns="http://www.bbraun.com/HC/AutoProgramming">
+# calculated_checksum_value
+# </Checksum>
+# <ChecksumTotal>
+# calculated_outer_checksum_value
+# </ChecksumTotal>
+#
+# An example of what goes out:
+#    <ProposalList xmlns="http://www.bbraun.com/HC/AutoProgramming">
+#    1234
+#    </Checksum>
+#    <ChecksumTotal>
+#    5678
+#    </ChecksumTotal>
+#
+#################################################################
+
+
+def define_autoprog_static(session):
+    dl_line_1 = String(
+        name="proposal_header", default_value='<ProposalList xmlns="http://www.mycompany.com/HC/AutoProgramming">'
+    )
+    # insert crc here
+    dl_line_9 = String(name="crc_end", default_value="</Checksum>", fuzzable=False)
+    dl_line_11 = String(name="outer_crc", default_value="<ChecksumTotal>", fuzzable=False)
+    # insert outer crc here
+    dl_line_12 = String(name="outer_crcend", default_value="</ChecksumTotal>", fuzzable=False)
+
+    reqW = Request("autoprog")
+    block = Block(name="autoprogB", request=reqW)
+    reqW.push(block)
+    crcValue = Checksum(
+        name="firstCRC16", block_name="autoprogB", request=reqW, algorithm=getCrc16Ccitt, length=16, fuzzable=False
+    )
+    crcValue_outer = Checksum(
+        name="CRC16_outer", block_name="autoprogB", request=reqW, algorithm=getCrc16Ccitt, length=16, fuzzable=False
+    )
+
+    block.push(dl_line_1)
+    reqW.push(crcValue)
+    block.push(dl_line_9)
+    block.push(dl_line_11)
+    reqW.push(crcValue_outer)
+    block.push(dl_line_12)
+    reqW.pop()
```

Comment:
Why use push/pop when we could define the protocol from top to bottom with children like it's done in the http_simple example?
https://github.com/jtpereyda/boofuzz/blob/71f69bf829f98114100d3c7815caa5e827980d1a/examples/http_simple.py#L22-L30

That would make it a bit easier to read I think.

Reply:
I like that approach better as well. @ptdropper this change can be optional for this PR.

Reply:
This is the style I have used in my example and in my working code. I found the other style difficult to use correctly. My example works so I am keeping the style.